atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java

```diff
@@ -222,23 +223,35 @@ private static void login(Client client, CassandraCredentialsConfig config) thro
      * This will check both ip address/hostname, and uses the IP address associated with the socket, rather
      * that what has been provided. Hostname/ip address are both need to be checked, as historically we've
      * connected to Cassandra directly using IP addresses, and therefore need to support such cases.
+     *
+     * Will only throw when throwOnFailure is true, even if the socket is closed during verification.
      */
     @VisibleForTesting
     static void verifyEndpoint(CassandraServer cassandraServer, SSLSocket socket, boolean throwOnFailure)
-            throws SafeSSLPeerUnverifiedException {
-        boolean endpointVerified = Stream.of(
-                        socket.getInetAddress().getHostAddress(), cassandraServer.cassandraHostName())
-                .anyMatch(address -> hostnameVerifier.verify(address, socket.getSession()));
-
+            throws SafeSSLPeerUnverifiedException, SocketException {
+        Set<String> endpointsToCheck = getEndpointsToCheck(cassandraServer, socket);
+        boolean endpointVerified =
+                endpointsToCheck.stream().anyMatch(address -> hostnameVerifier.verify(address, socket.getSession()));
+        if (socket.isClosed()) {
+            if (throwOnFailure) {
+                throw new SocketException("Unable to verify hostnames as socket is closed.");
```

Comment:
Wonder if we should wrap this exception in `SafeSSLPeerUnverifiedException`. 
Not too fussed either way, whichever way is more uniform in this class.

Reply:
Sounds good to me